step-automation-packages/step-automation-packages-yaml/src/main/java/step/automation/packages/yaml/AutomationPackageDescriptorReader.java

```diff
@@ -0,0 +1,143 @@
+        return readAutomationPackageYamlFile(yamlFragment, AutomationPackageFragmentYaml.class);
+    }
+
+    protected <T extends AutomationPackageFragmentYaml> T readAutomationPackageYamlFile(InputStream yaml, Class<T> targetClass) throws AutomationPackageReadingException {
+        try {
+            String yamlDescriptorString = new String(yaml.readAllBytes(), StandardCharsets.UTF_8);
+
+            if (jsonSchema != null) {
+                try {
+                    JsonSchemaValidator.validate(jsonSchema, yamlObjectMapper.readTree(yamlDescriptorString).toString());
+                } catch (Exception ex) {
+                    throw new YamlPlanValidationException(ex.getMessage(), ex);
+                }
+            }
+
+            T res = yamlObjectMapper.readValue(yamlDescriptorString, targetClass);
+
+            if (!res.getKeywords().isEmpty()) {
+                log.info("{} keyword(s) found in automation package", res.getKeywords().size());
+            }
+            if (!res.getPlans().isEmpty()) {
+                log.info("{} plan(s) found in automation package", res.getPlans().size());
+            }
+            if (!res.getScheduler().isEmpty()) {
+                log.info("{} scheduled task(s) found in automation package", res.getScheduler().size());
+            }
+            if (res.getFragments().isEmpty()) {
+                log.info("{} imported fragment(s) found in automation package", res.getFragments().size());
+            }
+            return res;
+        } catch (IOException | YamlPlanValidationException e) {
+            throw new AutomationPackageReadingException("Unable to read the automation package yaml", e);
+        }
+    }
+
+    protected String readJsonSchema(String jsonSchemaPath) {
+        try (InputStream jsonSchemaInputStream = this.getClass().getClassLoader().getResourceAsStream(jsonSchemaPath)) {
+            if (jsonSchemaInputStream == null) {
+                throw new IllegalStateException("Json schema not found: " + jsonSchemaPath);
+            }
+            return new String(jsonSchemaInputStream.readAllBytes(), StandardCharsets.UTF_8);
+        } catch (IOException e) {
+            throw new IllegalStateException("Unable to load json schema: " + jsonSchemaPath, e);
+        }
+    }
+
+    protected ObjectMapper createYamlObjectMapper() {
+        YAMLFactory yamlFactory = new YAMLFactory();
+
+        // Disable native type id to enable conversion to generic Documents
+        yamlFactory.disable(YAMLGenerator.Feature.USE_NATIVE_TYPE_ID);
+        ObjectMapper yamlMapper = DefaultJacksonMapperProvider.getObjectMapper(yamlFactory);
+
+        // configure custom deserializers
+        SimpleModule module = new SimpleModule();
+
+        // register serializers/deserializers to read yaml plans
+        // TODO: we don't want to use the default upgradable plan serializer, because the plan version is defined via automation package schema version, but not inside the plan...
```

Comment:
Yes. It's a good point for discussion, because for standalone plans we have the following mechanism for deserialization:
- check the actual yaml plans version against specified in yaml file
- try to upgrade (run migration tasks) the yaml to the actual yaml version
- run deserializer over the upgraded yaml

But when we're speaking about the plans inside the automation package, the yaml plan version is not defined inside the each plan, but is determined by the version of automation package. And theoretically we can receive the outdated yaml descriptor containing outdated plans, and in this case we want to upgrade the plans to the actual version...

Reply:
@iegorov777 As discussed during our call, the migration manager should be implemented for the complete automation package. I.e. it should be possible to implement migration tasks for all the entities contained in an automation package. This will be done after merging this PR